core/iwasm/libraries/wasi-nn/src/wasi_nn_onnx.cpp

```diff
@@ -0,0 +1,834 @@
+#include <string.h>
+#include <mutex>
+#include <vector>
+#include <unordered_map>
+#include "bh_platform.h"
+#include "wasi_nn_backend.h"
+#include "utils/logger.h"
+#include "onnxruntime_c_api.h"
+
+/* Maximum number of graphs and execution contexts */
+#define MAX_GRAPHS 4
+#define MAX_CONTEXTS 4
+
+/* Graph structure */
+typedef struct {
+    OrtSession *session;
+    bool is_initialized;
+} OnnxRuntimeGraph;
+
+/* Execution context structure */
+typedef struct {
+    OrtMemoryInfo *memory_info;
+    std::vector<const char *> input_names;
+    std::vector<const char *> output_names;
+    std::unordered_map<uint32_t, OrtValue *> inputs;
+    std::unordered_map<uint32_t, OrtValue *> outputs;
+    OnnxRuntimeGraph *graph;
+    bool is_initialized;
+} OnnxRuntimeExecCtx;
+
+/* ONNX Runtime context structure */
+typedef struct {
+    OrtEnv *env;
+    OrtSessionOptions *session_options;
+    OrtAllocator *allocator;
+    const OrtApi *ort_api;
+    std::mutex mutex;
+    bool is_initialized;
+    OnnxRuntimeGraph graphs[MAX_GRAPHS];
+    OnnxRuntimeExecCtx exec_ctxs[MAX_CONTEXTS];
+} OnnxRuntimeContext;
+
+/* Helper functions */
+static void
+check_status_and_log(const OnnxRuntimeContext *ctx, OrtStatus *status)
+{
+    if (status != nullptr) {
+        const char *msg = ctx->ort_api->GetErrorMessage(status);
+        NN_ERR_PRINTF("ONNX Runtime error: %s", msg);
+        ctx->ort_api->ReleaseStatus(status);
+    }
+}
+
+static wasi_nn_error
+convert_ort_error_to_wasi_nn_error(const OnnxRuntimeContext *ctx,
+                                   OrtStatus *status)
+{
+    if (status == nullptr) {
+        return success;
+    }
+
+    wasi_nn_error err;
+    OrtErrorCode code = ctx->ort_api->GetErrorCode(status);
+    const char *msg = ctx->ort_api->GetErrorMessage(status);
+
+    NN_ERR_PRINTF("ONNX Runtime error: %s", msg);
+
+    switch (code) {
+        case ORT_INVALID_ARGUMENT:
+            err = invalid_argument;
+            break;
+        case ORT_RUNTIME_EXCEPTION:
+            err = runtime_error;
+            break;
+        case ORT_NOT_IMPLEMENTED:
+            err = unsupported_operation;
+            break;
+        case ORT_INVALID_PROTOBUF:
+            err = invalid_encoding;
+            break;
+        case ORT_MODEL_LOADED:
+            err = too_large;
+            break;
+        case ORT_INVALID_GRAPH:
+            err = invalid_encoding;
+            break;
+        default:
+            err = runtime_error;
+            break;
+    }
+
+    ctx->ort_api->ReleaseStatus(status);
+    return err;
+}
+
+static bool
+convert_ort_type_to_wasi_nn_type(ONNXTensorElementDataType ort_type,
+                                 tensor_type *tensor_type)
+{
+    switch (ort_type) {
+        case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT:
+            *tensor_type = fp32;
+            break;
+        case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16:
+            *tensor_type = fp16;
+            break;
+#if WASM_ENABLE_WASI_EPHEMERAL_NN != 0
```

Comment:
do you still need the legacy abi (WASM_ENABLE_WASI_EPHEMERAL_NN == 0)?

we recently have dropped the legacy abi explicitly for the majority of backends: https://github.com/bytecodealliance/wasm-micro-runtime/pull/4468

the legacy abi has subtle differences, which this PR doesn't seem to implement.
cf. https://github.com/bytecodealliance/wasm-micro-runtime/blob/390e78030e53d2d2302ec24a26c847ea400693ad/core/iwasm/libraries/wasi-nn/src/wasi_nn_tensorflowlite.cpp#L430-L433

Reply:
after the "use WASM_ENABLE_WASI_EPHEMERAL_NN" commit, these WASM_ENABLE_WASI_EPHEMERAL_NN == 0 ifdef blocks are dead code. can you consider to remove them?